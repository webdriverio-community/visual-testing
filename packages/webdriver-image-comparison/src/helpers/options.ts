import { DEFAULT_FORMAT_STRING, DEFAULT_SHADOW, DEFAULT_TABBABLE_OPTIONS, FULL_PAGE_SCROLL_TIMEOUT } from './constants.js'
import type { ClassOptions, DefaultOptions } from './options.interfaces'
import { LogLevel } from './options.interfaces'
import type { MethodImageCompareCompareOptions, ScreenMethodImageCompareCompareOptions } from '../methods/images.interfaces'

/**
 * Determine the default options
 */
export function defaultOptions(options: ClassOptions): DefaultOptions {
    return {
        /**
         * Module options
         */
        addressBarShadowPadding: options.addressBarShadowPadding || DEFAULT_SHADOW.ADDRESS_BAR,
        autoElementScroll: Object.prototype.hasOwnProperty.call(options, 'autoElementScroll')
            ? Boolean(options.autoElementScroll)
            : true,
        addIOSBezelCorners: options.addIOSBezelCorners || false,
        autoSaveBaseline: options.autoSaveBaseline || false,
        clearFolder: options.clearRuntimeFolder || false,
        formatImageName: options.formatImageName || DEFAULT_FORMAT_STRING,
        isHybridApp: options.isHybridApp || false,
        logLevel: options.logLevel || LogLevel.info,
        savePerInstance: options.savePerInstance || false,
        toolBarShadowPadding: options.toolBarShadowPadding || DEFAULT_SHADOW.TOOL_BAR,

        /**
     * Module and method options
     */
        disableCSSAnimation: options.disableCSSAnimation || false,
        fullPageScrollTimeout: options.fullPageScrollTimeout || FULL_PAGE_SCROLL_TIMEOUT,
        hideScrollBars: Object.prototype.hasOwnProperty.call(options, 'hideScrollBars')
            ? Boolean(options.hideScrollBars)
            : true,

        /**
     * Compare options
     */
        compareOptions: {
            blockOutSideBar: !!options.blockOutSideBar,
            blockOutStatusBar: !!options.blockOutStatusBar,
            blockOutToolBar: !!options.blockOutToolBar,
            ignoreAlpha: options.ignoreAlpha || false,
            ignoreAntialiasing: options.ignoreAntialiasing || false,
            ignoreColors: options.ignoreColors || false,
            ignoreLess: options.ignoreLess || false,
            ignoreNothing: options.ignoreNothing || false,
            rawMisMatchPercentage: options.rawMisMatchPercentage || false,
            returnAllCompareData: options.returnAllCompareData || false,
            saveAboveTolerance: options.saveAboveTolerance || 0,
            scaleImagesToSameSize: options.scaleImagesToSameSize || false,
        },

        /**
     * Tabbable options
     */
        tabbableOptions: {
            circle: {
                ...DEFAULT_TABBABLE_OPTIONS.circle,
                ...(options.tabbableOptions && options.tabbableOptions.circle ? options.tabbableOptions.circle : {}),
            },
            line: {
                ...DEFAULT_TABBABLE_OPTIONS.line,
                ...(options.tabbableOptions && options.tabbableOptions.line ? options.tabbableOptions.line : {}),
            },
        },
    }
}

/**
 * Determine the screen method compare options
 */
export function screenMethodCompareOptions(
    options: ScreenMethodImageCompareCompareOptions,
): ScreenMethodImageCompareCompareOptions {
    return {
        ...('blockOutSideBar' in options ? { blockOutSideBar: options.blockOutSideBar } : {}),
        ...('blockOutStatusBar' in options ? { blockOutStatusBar: options.blockOutStatusBar } : {}),
        ...('blockOutToolBar' in options ? { blockOutToolBar: options.blockOutToolBar } : {}),
        ...methodCompareOptions(options),
    }
}

/**
 * Determine the method compare options
 */
export function methodCompareOptions(options: any): MethodImageCompareCompareOptions {
    return {
        ...('blockOut' in options ? { blockOut: options.blockOut } : {}),
        ...('ignoreAlpha' in options ? { ignoreAlpha: options.ignoreAlpha } : {}),
        ...('ignoreAntialiasing' in options ? { ignoreAntialiasing: options.ignoreAntialiasing } : {}),
        ...('ignoreColors' in options ? { ignoreColors: options.ignoreColors } : {}),
        ...('ignoreLess' in options ? { ignoreLess: options.ignoreLess } : {}),
        ...('ignoreNothing' in options ? { ignoreNothing: options.ignoreNothing } : {}),
        ...('rawMisMatchPercentage' in options ? { rawMisMatchPercentage: options.rawMisMatchPercentage } : {}),
        ...('returnAllCompareData' in options ? { returnAllCompareData: options.returnAllCompareData } : {}),
        ...('saveAboveTolerance' in options ? { saveAboveTolerance: options.saveAboveTolerance } : {}),
        ...('scaleImagesToSameSize' in options ? { scaleImagesToSameSize: options.scaleImagesToSameSize } : {}),
    }
}
