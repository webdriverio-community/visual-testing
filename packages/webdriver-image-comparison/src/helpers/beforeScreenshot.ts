import hideScrollBars from '../clientSideScripts/hideScrollbars.js'
import setCustomCss from '../clientSideScripts/setCustomCss.js'
import { CUSTOM_CSS_ID } from './constants.js'
import { checkIsMobile, getAddressBarShadowPadding, getToolBarShadowPadding, waitFor } from './utils.js'
import getEnrichedInstanceData from '../methods/instanceData.js'
import type { BeforeScreenshotOptions, BeforeScreenshotResult } from './beforeScreenshot.interfaces'
import type { Executor } from '../methods/methods.interfaces'
import hideRemoveElements from '../clientSideScripts/hideRemoveElements.js'
import { LogLevel } from './options.interfaces'

/**
 * Methods that need to be executed before a screenshot will be taken
 */
export default async function beforeScreenshot(
    executor: Executor,
    options: BeforeScreenshotOptions,
    addShadowPadding = false,
): Promise<BeforeScreenshotResult> {
    const { browserName, nativeWebScreenshot, platformName } = options.instanceData
    const {
        addressBarShadowPadding,
        disableCSSAnimation,
        hideElements,
        logLevel,
        noScrollBars,
        removeElements,
        toolBarShadowPadding,
    } = options
    const addressBarPadding = getAddressBarShadowPadding({
        platformName,
        browserName,
        nativeWebScreenshot,
        addressBarShadowPadding,
        addShadowPadding,
    })
    const toolBarPadding = getToolBarShadowPadding({ platformName, browserName, toolBarShadowPadding, addShadowPadding })

    // Hide the scrollbars
    if (noScrollBars) {
        await executor(hideScrollBars, noScrollBars)
    }

    // Hide and or Remove elements
    if (hideElements.length > 0 || removeElements.length > 0) {
        try {
            await executor(hideRemoveElements, { hide: hideElements, remove: removeElements }, true)
        } catch (e) {
            if (logLevel === LogLevel.debug || logLevel === LogLevel.warn) {
                console.log(
                    '\x1b[33m%s\x1b[0m',
                    `
#####################################################################################
 WARNING:
 (One of) the elements that needed to be hidden or removed could not be found on the
 page and caused this error
 Error: ${e}
 We made sure the test didn't break.
#####################################################################################
`,
                )
            }
        }
    }

    // Set some custom css
    if (disableCSSAnimation || checkIsMobile(platformName)) {
        await executor(setCustomCss, { addressBarPadding, disableCSSAnimation, id: CUSTOM_CSS_ID, toolBarPadding })
        // Wait at least 500 milliseconds to make sure the css is applied
        // Not every device is fast enough to apply the css faster
        await waitFor(500)
    }

    // Get all the needed instance data
    const instanceOptions = {
        addressBarShadowPadding: options.addressBarShadowPadding,
        toolBarShadowPadding: options.toolBarShadowPadding,
        ...options.instanceData,
    }

    return getEnrichedInstanceData(executor, instanceOptions, addShadowPadding)
}
