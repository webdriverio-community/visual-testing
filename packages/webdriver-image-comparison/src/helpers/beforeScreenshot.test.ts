import { describe, it, expect, vi } from 'vitest'
import beforeScreenshot from './beforeScreenshot.js'
import { LogLevel } from './options.interfaces'

describe('beforeScreenshot', () => {
    it('should be able to return the enriched instance data with default options', async () => {
        const MOCKED_EXECUTOR = vi.fn().mockReturnValue('')

        const options = {
            instanceData: {
                browserName: 'browserName',
                browserVersion: 'browserVersion',
                deviceName: 'deviceName',
                logName: 'logName',
                name: 'name',
                nativeWebScreenshot: false,
                platformName: 'platformName',
                platformVersion: 'platformVersion',
            },
            addressBarShadowPadding: 6,
            disableCSSAnimation: true,
            logLevel: LogLevel.debug,
            noScrollBars: true,
            toolBarShadowPadding: 6,
            hideElements: [<HTMLElement>(<any>'<div></div>')],
            removeElements: [<HTMLElement>(<any>'<div></div>')],
        }

        expect(await beforeScreenshot(MOCKED_EXECUTOR, options)).toMatchSnapshot()
    })

    it('should be able to return the enriched instance data with `addShadowPadding: true`', async () => {
        const MOCKED_EXECUTOR = vi.fn().mockReturnValue('')

        const options = {
            instanceData: {
                browserName: 'browserName',
                browserVersion: 'browserVersion',
                deviceName: 'deviceName',
                logName: 'logName',
                name: 'name',
                nativeWebScreenshot: false,
                platformName: 'platformName',
                platformVersion: 'platformVersion',
            },
            addressBarShadowPadding: 6,
            disableCSSAnimation: true,
            noScrollBars: true,
            logLevel: LogLevel.debug,
            toolBarShadowPadding: 6,
            hideElements: [<HTMLElement>(<any>'<div></div>')],
            removeElements: [<HTMLElement>(<any>'<div></div>')],
        }

        expect(await beforeScreenshot(MOCKED_EXECUTOR, options, true)).toMatchSnapshot()
    })
})
