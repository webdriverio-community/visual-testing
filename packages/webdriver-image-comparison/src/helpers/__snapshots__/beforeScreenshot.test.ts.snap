// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`beforeScreenshot > should be able to return the enriched instance data with \`addShadowPadding: true\` 1`] = `
{
  "addressBarShadowPadding": 0,
  "appName": "appName",
  "browserName": "browserName",
  "browserVersion": "browserVersion",
  "deviceName": "deviceName",
  "devicePixelRatio": 1,
  "deviceScreenSize": {
    "height": 1,
    "width": 1,
  },
  "isAndroid": false,
  "isAndroidChromeDriverScreenshot": false,
  "isAndroidNativeWebScreenshot": false,
  "isIOS": false,
  "isMobile": false,
  "isTestInBrowser": true,
  "isTestInMobileBrowser": false,
  "logName": "logName",
  "name": "name",
  "nativeWebScreenshot": false,
  "platformName": "platformName",
  "platformVersion": "platformVersion",
  "toolBarShadowPadding": 0,
}
`;

exports[`beforeScreenshot > should be able to return the enriched instance data with default options 1`] = `
{
  "addressBarShadowPadding": 0,
  "appName": "appName",
  "browserName": "browserName",
  "browserVersion": "browserVersion",
  "deviceName": "deviceName",
  "devicePixelRatio": 1,
  "deviceScreenSize": {
    "height": 1,
    "width": 1,
  },
  "isAndroid": false,
  "isAndroidChromeDriverScreenshot": false,
  "isAndroidNativeWebScreenshot": false,
  "isIOS": false,
  "isMobile": false,
  "isTestInBrowser": true,
  "isTestInMobileBrowser": false,
  "logName": "logName",
  "name": "name",
  "nativeWebScreenshot": false,
  "platformName": "platformName",
  "platformVersion": "platformVersion",
  "toolBarShadowPadding": 0,
}
`;
